File: vet/cliente/src/VacinaPerecivel.cpp
#include "../include/VacinaPerecivel.hpp"

VacinaPerecivel::VacinaPerecivel(){};

VacinaPerecivel::VacinaPerecivel(int id,const string& nome,double preco, const string& fabricante,
        const string& regM, const string& espA, const string& vAdm,
        const string& tipoAgente, const string& sorotipo, int numDoses, 
        string dataValidade, string requisitoArmazenamento, double temperaturaMinima, double temperaturaMaxima)
    : ProdutoBiologico(id, nome, preco, fabricante, regM, espA, vAdm, tipoAgente, sorotipo, numDoses),
    dataValidade(dataValidade), requisitoArmazenamento(requisitoArmazenamento), temperaturaMinima(temperaturaMinima), temperaturaMaxima(temperaturaMaxima) {}


const string& VacinaPerecivel::getDataValidade() const{
    return this->dataValidade;
}

bool VacinaPerecivel::ehVencido() const{
    int dia, mes, ano;
    //quebrando a string em data
    sscanf(this->dataValidade.c_str(), "%d/%d/%d", &dia, &mes, &ano);

    //struct da validade
    tm validade = {};
    validade.tm_mday = dia;
    validade.tm_mon = mes - 1; //mes de 0 a 11
    validade.tm_year = ano - 1900;

    time_t tempoValidade = mktime(&validade);  // data -> numero(timestamp)

    //data atual
    time_t agora = time(nullptr);

    return difftime(agora, tempoValidade) > 0;
} 

void VacinaPerecivel::setDataValidade(const string& dataValidade){
    this->dataValidade = dataValidade;
}

string VacinaPerecivel::getRequisitoArmazenamento() const{
    return this->requisitoArmazenamento;
}

void VacinaPerecivel::setRequisitoArmazenamento(const string& requisitoArmazenamento){
    this->requisitoArmazenamento = requisitoArmazenamento;
}

double VacinaPerecivel::getTemperaturaMaxima() const{
    return this->temperaturaMaxima;
}

void VacinaPerecivel::setTemperaturaMaxima(double temperaturaMaxima){
    this->temperaturaMaxima = temperaturaMaxima;
}

double VacinaPerecivel::getTemperaturaMinima() const{
    return this->temperaturaMinima;
}

void VacinaPerecivel::setTemperaturaMinima(double temperaturaMinima){
    this->temperaturaMinima = temperaturaMinima;
}

string VacinaPerecivel::toString() const{
    return "VacinaPerecivel{id=" + to_string(this->getId()) +
    ", nome='" + this->getNome() + 
    "', validade='" + this->dataValidade + 
    "', vencida=" + (this->ehVencido() ? "true" : "false") + 
    ", armazenamento='" + this->requisitoArmazenamento + 
    "', temp=[" + to_string(this->temperaturaMinima) + 
    ";" + to_string(this->temperaturaMaxima) + "]°C}";

}

File: vet/src/pojo/VacinaPerecivel.java
package pojo;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

/**
 * POJO 5 — VacinaPerecivel
 * Subclasse de ProdutoBiologico. Implementa a interface Perecivel.
 * Representa vacinas que requerem cadeia de frio e têm validade.
 */
public class VacinaPerecivel extends ProdutoBiologico implements Perecivel {

    private String dataValidade;           // formato dd/MM/yyyy
    private String requisitoArmazenamento; // Ex: "Refrigerado 2-8°C"
    private double temperaturaMinima;      // °C
    private double temperaturaMaxima;      // °C

    public VacinaPerecivel() {}

    public VacinaPerecivel(int id, String nome, double preco, String fabricante,
                           String registroMapa, String especieAlvo, String viaAdministracao,
                           String tipoAgente, String sorotipo, int numDoses,
                           String dataValidade, String requisitoArmazenamento,
                           double temperaturaMinima, double temperaturaMaxima) {
        super(id, nome, preco, fabricante, registroMapa, especieAlvo, viaAdministracao,
              tipoAgente, sorotipo, numDoses);
        this.dataValidade           = dataValidade;
        this.requisitoArmazenamento = requisitoArmazenamento;
        this.temperaturaMinima      = temperaturaMinima;
        this.temperaturaMaxima      = temperaturaMaxima;
    }

    // ── Implementação de Perecivel ───────────────────────────────────────────

    @Override
    public String getDataValidade() { return dataValidade; }

    @Override
    public boolean isVencido() {
        DateTimeFormatter fmt = DateTimeFormatter.ofPattern("dd/MM/yyyy");
        LocalDate validade    = LocalDate.parse(dataValidade, fmt);
        return LocalDate.now().isAfter(validade);
    }

    @Override
    public String getRequisitoArmazenamento() { return requisitoArmazenamento; }

    // ── Getters / Setters extras ─────────────────────────────────────────────

    public void   setDataValidade(String dataValidade)                   { this.dataValidade = dataValidade; }
    public void   setRequisitoArmazenamento(String r)                    { this.requisitoArmazenamento = r; }

    public double getTemperaturaMinima()                                  { return temperaturaMinima; }
    public void   setTemperaturaMinima(double t)                          { this.temperaturaMinima = t; }

    public double getTemperaturaMaxima()                                  { return temperaturaMaxima; }
    public void   setTemperaturaMaxima(double t)                          { this.temperaturaMaxima = t; }

    @Override
    public String toString() {
        return "VacinaPerecivel{" +
               "id=" + getId() +
               ", nome='" + getNome() + '\'' +
               ", validade='" + dataValidade + '\'' +
               ", vencida=" + isVencido() +
               ", armazenamento='" + requisitoArmazenamento + '\'' +
               ", temp=[" + temperaturaMinima + ";" + temperaturaMaxima + "]°C}";
    }
}